src/BlockTypes/SingleProduct.php

```diff
@@ -0,0 +1,153 @@
+
+		$html = sprintf(
+			'<div class="wp-block-woocommerce-single-product %1$s">
+				%2$s
+			</div>',
+			esc_attr( $classname ),
+			$content
+		);
+
+		return $html;
+	}
+
+	/**
+	 * Update the context by injecting the correct post data
+	 * for each one of the Single Product inner blocks.
+	 *
+	 * @param array    $context Block context.
+	 * @param array    $block Block attributes.
+	 * @param WP_Block $parent_block Block instance.
+	 *
+	 * @return array Updated block context.
+	 */
+	public function update_context( $context, $block, $parent_block ) {
+		if ( 'woocommerce/single-product' == $block['blockName']
+			&& isset( $block['attrs']['productId'] ) ) {
+				$this->product_id = $block['attrs']['productId'];
+
+				$this->single_product_inner_blocks_names = array_reverse(
+					$this->extract_single_product_inner_block_names( $block )
+				);
+		}
+
+		$this->replace_post_for_single_product_inner_block( $block, $context );
+
+		return $context;
+	}
+
+	/**
+	 * Extract the inner block names for the Single Product block. This way it's possible
+	 * to map all the inner blocks for a Single Product block and manipulate the data as needed.
+	 *
+	 * @param array $block The Single Product block or its inner blocks.
+	 * @param array $result Array of inner block names.
+	 *
+	 * @return array Array containing all the inner block names of a Single Product block.
+	 */
+	protected function extract_single_product_inner_block_names( $block, &$result = [] ) {
+		if ( isset( $block['blockName'] ) ) {
+			$result[] = $block['blockName'];
+		}
+
+		if ( isset( $block['innerBlocks'] ) ) {
+			foreach ( $block['innerBlocks'] as $inner_block ) {
+				$this->extract_single_product_inner_block_names( $inner_block, $result );
+			}
+		}
+		return $result;
+	}
+
+	/**
+	 * Replace the global post for the Single Product inner blocks and reset it after.
+	 *
+	 * This is needed because some of the inner blocks may use the global post
+	 * instead of fetching the product through the `productId` attribute, so even if the
+	 * `productId` is passed to the inner block, it will still use the global post.
+	 *
+	 * @param array $block Block attributes.
+	 * @param array $context Block context.
+	 */
+	protected function replace_post_for_single_product_inner_block( $block, &$context ) {
+		if ( $this->single_product_inner_blocks_names ) {
+			$block_name = array_pop( $this->single_product_inner_blocks_names );
+
+			if ( $block_name === $block['blockName'] ) {
+				// @todo This is a temporary fix to make the Post Excerpt block work while https://github.com/WordPress/gutenberg/pull/49495 is not merged
+				if ( 'core/post-excerpt' === $block_name ) {
+					global $post;
+					// phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+					$post = get_post( $this->product_id );
+					setup_postdata( $post );
+				}
+				$context['postId'] = $this->product_id;
+				if ( ! $this->single_product_inner_blocks_names ) {
```

Comment:
Do you mind clarifying the logic here? This condition is currently inside another `if ( $this->single_product_inner_blocks_names )` so I don't think there will ever be a case where this condition will be true.

Reply:
Sure! The idea is that I'm using `array_pop` on line 125 to keep getting each Inner Block name until the array is empty. The `render_block_context` is called once for each block in a sequential way, so after iterating over all the inner blocks, the array should be empty, and then the `wp_reset_postdata` is called.

Now that you mention I think it could be better if we move this condition outside the `if ( $block_name === $block['blockName'] ) ` so we can leave it at the same level of the `array_pop`. I just did this change!

Reply:
> Sure! The idea is that I'm using array_pop on line 125 to keep getting each Inner Block name until the array is empty. The render_block_context is called once for each block in a sequential way, so after iterating over all the inner blocks, the array should be empty, and then the wp_reset_postdata is called.

> Now that you mention I think it could be better if we move this condition outside the if ( $block_name === $block['blockName'] )  so we can leave it at the same level of the array_pop. I just did this change!

Thanks for clarifying & the update! Perhaps running `wp_reset_postdata()` by the end of every render of this block could be avoided, but since it's behind the feature flag and is also temporary, all good 👍